File: src/main/java/net/gtaun/shoebill/common/dialog/ListDialogItem.java
/**
 * Copyright (C) 2012-2014 MK124
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.gtaun.shoebill.common.dialog;

import java.util.function.BooleanSupplier;
import java.util.function.Supplier;

public class ListDialogItem
{
	@SuppressWarnings("unchecked")
	public static abstract class AbstractItemBuilder
	<ItemType extends ListDialogItem, ItemBuilderType extends AbstractItemBuilder<ItemType, ItemBuilderType>>
	{
		protected final ItemType item;
		
		
		protected AbstractItemBuilder(ItemType item)
		{
			this.item = item;
		}
		
		public ItemBuilderType data(Object data)
		{
			item.setData(data);
			return (ItemBuilderType) this;
		}
		
		public ItemBuilderType itemText(String itemText)
		{
			item.setItemText(itemText);
			return (ItemBuilderType) this;
		}

		public ItemBuilderType itemText(Supplier<String> textSupplier)
		{
			item.setItemText(textSupplier);
			return (ItemBuilderType) this;
		}
		
		public ItemBuilderType itemText(ItemTextSupplier<?> textSupplier)
		{
			item.setItemText(textSupplier);
			return (ItemBuilderType) this;
		}

		public ItemBuilderType enabled(BooleanSupplier enabledSupplier)
		{
			item.setItemEnabledSupplier(enabledSupplier);
			return (ItemBuilderType) this;
		}
		
		public ItemBuilderType enabled(ItemBooleanSupplier<?> enabledSupplier)
		{
			item.setItemEnabledSupplier(enabledSupplier);
			return (ItemBuilderType) this;
		}
		
		public ItemBuilderType onSelect(ItemSelectSimpleHandler selectHandler)
		{
			item.setSelectHandler(selectHandler);
			return (ItemBuilderType) this;
		}
		
		public ItemBuilderType onSelect(ItemSelectHandler<?> selectHandler)
		{
			item.setSelectHandler(selectHandler);
			return (ItemBuilderType) this;
		}
		
		public ItemType build()
		{
			return item;
		}
	}
	
	public static class ItemBuilder extends AbstractItemBuilder<ListDialogItem, ItemBuilder>
	{
		private ItemBuilder()
		{
			super(new ListDialogItem("-"));
		}
	}
	
	public static AbstractItemBuilder<?, ?> create()
	{
		return new ItemBuilder();
	}
	
	
	@FunctionalInterface
	public interface ItemTextSupplier<ItemData>
	{
		String getItemText(ItemData data);
	}
	
	@FunctionalInterface
	public interface ItemBooleanSupplier<ItemData>
	{
		boolean getBoolean(ItemData data);
	}

	@FunctionalInterface
	public interface ItemSelectSimpleHandler
	{
		void onItemSelect(ListDialogItem item);
	}
	
	@FunctionalInterface
	public interface ItemSelectHandler<ItemData>
	{
		void onItemSelect(ListDialogItem item, ItemData data);
	}
	
	
	protected ListDialog currentDialog;
	
	protected Object data;
	protected ItemTextSupplier<Object> itemTextSupplier;
	protected ItemBooleanSupplier<Object> itemEnabledSupplier;
	protected ItemSelectHandler<Object> selectHandler;
	

	protected <DataType> ListDialogItem(String itemText)
	{
		setItemText(itemText);
	}
	
	protected <DataType> ListDialogItem(Supplier<String> textSupplier)
	{
		setItemText(textSupplier);
	}
	
	public ListDialogItem(String itemText, ItemSelectSimpleHandler handler)
	{
		setItemText(itemText);
		setSelectHandler(handler);
	}
	
	public ListDialogItem(Supplier<String> textSupplier, ItemSelectSimpleHandler handler)
	{
		setItemText(textSupplier);
		setSelectHandler(handler);
	}

	public ListDialogItem(String itemText, BooleanSupplier enabledSupplier, ItemSelectSimpleHandler handler)
	{
		setItemText(itemText);
		setItemEnabledSupplier(enabledSupplier);
		setSelectHandler(handler);
	}
	
	public ListDialogItem(Supplier<String> textSupplier, BooleanSupplier enabledSupplier, ItemSelectSimpleHandler handler)
	{
		setItemText(textSupplier);
		setItemEnabledSupplier(enabledSupplier);
		setSelectHandler(handler);
	}
	
	public <DataType> ListDialogItem(DataType data, String itemText, ItemSelectHandler<DataType> handler)
	{
		setData(data);
		setItemText(itemText);
		setSelectHandler(handler);
	}
	
	public <DataType> ListDialogItem(DataType data, ItemTextSupplier<DataType> textSupplier, ItemSelectHandler<DataType> handler)
	{
		setData(data);
		setItemText(textSupplier);
		setSelectHandler(handler);
	}
	
	public <DataType> ListDialogItem(DataType data, String itemText, ItemBooleanSupplier<DataType> enabledSupplier, ItemSelectHandler<DataType> handler)
	{
		setData(data);
		setItemText(itemText);
		setItemEnabledSupplier(enabledSupplier);
		setSelectHandler(handler);
	}
	
	public <DataType> ListDialogItem(DataType data, ItemTextSupplier<DataType> textSupplier, ItemBooleanSupplier<DataType> enabledSupplier, ItemSelectHandler<DataType> handler)
	{
		setData(data);
		setItemText(textSupplier);
		setItemEnabledSupplier(enabledSupplier);
		setSelectHandler(handler);
	}
	
	public ListDialog getCurrentDialog()
	{
		return currentDialog;
	}

	public void setData(Object data)
	{
		this.data = data;
	}
	
	public void setItemText(String itemText)
	{
		itemTextSupplier = (d) -> itemText;
	}

	public void setItemText(Supplier<String> textSupplier)
	{
		itemTextSupplier = (d) -> textSupplier.get();
	}
	
	@SuppressWarnings("unchecked")
	public void setItemText(ItemTextSupplier<?> textSupplier)
	{
		itemTextSupplier = (ItemTextSupplier<Object>) textSupplier;
	}
	
	public void setItemEnabledSupplier(BooleanSupplier enabledSupplier)
	{
		itemEnabledSupplier = (d) -> enabledSupplier.getAsBoolean();
	}
	
	@SuppressWarnings("unchecked")
	public void setItemEnabledSupplier(ItemBooleanSupplier<?> enabledSupplier)
	{
		itemEnabledSupplier = (ItemBooleanSupplier<Object>) enabledSupplier;
	}

	public void setSelectHandler(ItemSelectSimpleHandler selectHandler)
	{
		this.selectHandler = (i, d) -> selectHandler.onItemSelect(i);
	}
	
	@SuppressWarnings("unchecked")
	public void setSelectHandler(ItemSelectHandler<?> selectHandler)
	{
		this.selectHandler = (ItemSelectHandler<Object>) selectHandler;
	}
	
	public String getItemText()
	{
		return itemTextSupplier != null ? itemTextSupplier.getItemText(null) : "-";
	}
	
	public boolean isEnabled()
	{
		if (itemEnabledSupplier == null) return true;
		return itemEnabledSupplier.getBoolean(data);
	}
	
	public void onItemSelect()
	{
		if (selectHandler != null) selectHandler.onItemSelect(this, data);
	}
}

File: src/main/java/net/gtaun/shoebill/common/dialog/ListDialogItem.kt
/**
 * Copyright (C) 2012-2014 MK124

 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at

 * http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.gtaun.shoebill.common.dialog

import net.gtaun.shoebill.common.AllOpen
import net.gtaun.shoebill.entities.Player
import java.util.function.BiConsumer

@AllOpen
class ListDialogItem {

    @AllOpen
    abstract class AbstractItemBuilder<T : ListDialogItem, B : AbstractItemBuilder<T, B>> {

        protected lateinit var item: T

        fun init(init: B.(B) -> Unit): B {
            init(this as B, this)
            return this
        }

        fun text(text: String) = text { text }
        fun textSupplier(supplier: DialogTextSupplier) = textSupplier { supplier }
        fun enabled(enabled: Boolean) = enabled { enabled }
        fun enabledSupplier(supplier: ItemBooleanSupplier) = enabledSupplier { supplier }

        fun onSelect(handler: BiConsumer<ListDialogItem, Player>) = onSelect {
            item, player ->
            handler.accept(item, player)
        }

        fun onSelectHandler(handler: ItemSelectHandler) = onSelectHandler { handler }

        fun text(init: B.() -> String): B {
            item.itemText = init(this as B)
            return this
        }

        fun textSupplier(init: B.() -> DialogTextSupplier): B {
            item.itemTextSupplier = init(this as B)
            return this
        }

        fun enabled(init: B.() -> Boolean): B {
            item.itemEnabledSupplier = ItemBooleanSupplier { init(this as B) }
            return this as B
        }

        fun enabledSupplier(init: B.() -> ItemBooleanSupplier): B {
            item.itemEnabledSupplier = init(this as B)
            return this
        }

        fun onSelect(init: (ListDialogItem, Player) -> Unit): B {
            item.selectHandler = ItemSelectHandler { item, player -> init(item, player) }
            return this as B
        }

        fun onSelectHandler(init: B.() -> ItemSelectHandler): B {
            item.selectHandler = init(this as B)
            return this
        }

        fun build(): T = item
    }

    @AllOpen
    class ListDialogItemBuilder : AbstractItemBuilder<ListDialogItem, ListDialogItemBuilder>() {
        init {
            item = ListDialogItem()
        }
    }

    var currentDialog: ListDialog? = null
        set

    final var itemTextSupplier: DialogTextSupplier? = null
    final var itemEnabledSupplier: ItemBooleanSupplier? = null
    final var selectHandler: ItemSelectHandler? = null

    constructor()

    @JvmOverloads
    constructor(text: String, enabledSupplier: ItemBooleanSupplier? = null,
                handler: ItemSelectHandler? = null) {
        itemTextSupplier = DialogTextSupplier { text }
        itemEnabledSupplier = enabledSupplier
        selectHandler = handler
    }

    @JvmOverloads
    constructor(textSupplier: DialogTextSupplier, enabledSupplier: ItemBooleanSupplier? = null,
                handler: ItemSelectHandler? = null) {
        itemTextSupplier = textSupplier
        itemEnabledSupplier = enabledSupplier
        selectHandler = handler
    }

    var itemText: String
        get() = itemTextSupplier?.get(currentDialog) ?: "-"
        set(itemText) {
            itemTextSupplier = DialogTextSupplier { itemText }
        }

    val isEnabled: Boolean
        get() = itemEnabledSupplier == null || itemEnabledSupplier?.get() ?: true

    fun onItemSelect(player: Player) {
        selectHandler?.onItemSelect(this, player) ?: return Unit
    }

    @FunctionalInterface
    interface ItemBooleanSupplier {
        fun get(): Boolean
    }

    @FunctionalInterface
    interface ItemSelectHandler {
        fun onItemSelect(item: ListDialogItem, player: Player)
    }

    companion object {

        @JvmStatic
        fun create() = ListDialogItemBuilder()

        fun ItemBooleanSupplier(handler: () -> Boolean) = object : ItemBooleanSupplier {
            override fun get(): Boolean {
                return handler()
            }
        }

        fun ItemSelectHandler(handler: (ListDialogItem, Player) -> Unit) = object : ItemSelectHandler {
            override fun onItemSelect(item: ListDialogItem, player: Player) {
                return handler(item, player)
            }
        }
    }
}
